File: haskell/doGuessing.hs

doGuessing :: Integer -> IO ()
doGuessing num = do
  putStrLn "Enter your guess:"
  guess <- getLine
  if read guess < num
    then do putStrLn "Too low!"
            doGuessing num
    else if read guess > num
            then do putStrLn "too high!"
                    doGuessing num
            else putStrLn "You win!"

doGuessing' :: Integer -> IO ()
doGuessing' num = do
  putStrLn "Enter your guess:"
  guess <- getLine
  case compare (read guess) num of
    LT  -> do putStrLn  "Too low!"
              doGuessing' num
    GT  -> do putStrLn "Too high!"
              doGuessing' num 
    EQ  -> do putStrLn "You win!"

File: haskell/doGuessing.c
#include <stdio.h>
#include <readline/readline.h>
#include <stdlib.h>

void doGuessing(int num) {
  printf("Enter your guess:\n");
  int guess = atoi(readline("$>"));
  if(guess == num) {
    printf("You win!\n");
    return ;
  }

  if(guess < num) {
    printf("Too low!\n");
    doGuessing(num);
  } else {
    printf("Too high!\n");
    doGuessing(num);
  }
}

int main() {
  doGuessing(23);
  return 0;
}

